### PR TITLE
Update balance helpers v0.6

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #2307 

---

# Add helper functions to balance change

includes:
get_balance
set_balance
increase_balance
decrease_balance

**Write why you are making the changes in this pull request**
to comply with 0.6 spec

**Write a summary of the changes you are making**
changes in:
- beacon-chain/core/helpers/rewards_penalties_test.go 
- beacon-chain/core/helpers/rewards_penalties.go
**Link anything that would be helpful or relevant to the reviewers**
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#get_balance

### CODE REVIEW

**Comment by terenc3t** [comment]
updated base to spec-v0.6

**Comment by prestonvanloon** [comment]
Please fix build issues

**Comment by shayzluf** [comment]
its the proto changes merge

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 1]
```diff
@@ -83,3 +83,67 @@ func InactivityPenalty(
 	validatorBalance := EffectiveBalance(state, validatorIndex)
 	return baseReward + validatorBalance*epochsSinceFinality/params.BeaconConfig().InactivityPenaltyQuotient/2
 }
+
+// GetBalance returns the Gwei balance of a validator by its index
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]: Ignore `Get` here

```suggestion
// Balance returns the Gwei balance of a validator by its index.
```

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 2]
```diff
@@ -83,3 +83,67 @@ func InactivityPenalty(
...
+//   def get_balance(state: BeaconState, index: ValidatorIndex) -> Gwei:
+//     """
+//     Return the balance for a validator with the given ``index``.
+//     """
+//     return state.balances[index]
+func GetBalance(state *pb.BeaconState, validatorIndex uint64) uint64 {
+	return state.ValidatorBalances[validatorIndex]
+}
+
+// SetBalance sets the Gwei balance of a validator by its index
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]:
```suggestion
// SetBalance sets the Gwei balance of a validator by its index.
```

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 3]
```diff
@@ -83,3 +83,67 @@ func InactivityPenalty(
...
+
+// GetBalance returns the Gwei balance of a validator by its index
+//
+// Spec pseudocode definition:
+//   def get_balance(state: BeaconState, index: ValidatorIndex) -> Gwei:
+//     """
+//     Return the balance for a validator with the given ``index``.
+//     """
+//     return state.balances[index]
+func GetBalance(state *pb.BeaconState, validatorIndex uint64) uint64 {
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]: `validatorIndex` feels too verbose here, prefer `state.ValidatorBalances[idx]` over `state.ValidatorBalances[validatorIndex]`

```suggestion
func GetBalance(state *pb.BeaconState, idx uint64) uint64 {
```

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 4]
```diff
@@ -83,3 +83,67 @@ func InactivityPenalty(
...
+//     """
+//     Set the balance for a validator with the given ``index`` in both ``BeaconState``
+//     and validator's rounded balance ``high_balance``.
+//     """
+//     validator = state.validator_registry[index]
+//     HALF_INCREMENT = HIGH_BALANCE_INCREMENT // 2
+//     if validator.high_balance > balance or validator.high_balance + 3 * HALF_INCREMENT < balance:
+//         validator.high_balance = balance - balance % HIGH_BALANCE_INCREMENT
+//     state.balances[index] = balance
+func SetBalance(state *pb.BeaconState, validatorIndex uint64, balance uint64) {
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]: Same, I think we'd prefer `state.ValidatorRegistry[idx]` over `state.ValidatorRegistry[validatorIndex]`

```suggestion
func SetBalance(state *pb.BeaconState, idx uint64, balance uint64) {
```

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 5]
```diff
@@ -83,3 +83,67 @@ func InactivityPenalty(
...
+	validator := state.ValidatorRegistry[validatorIndex]
+	halfIncrement := params.BeaconConfig().HighBalanceIncrement / 2
+	if validator.HighBalance > balance || validator.HighBalance+3*halfIncrement < balance {
+		validator.HighBalance = balance - balance%params.BeaconConfig().HighBalanceIncrement
+	}
+	state.ValidatorBalances[validatorIndex] = balance
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]: I think you need to return the state at the end

Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: I don't know. they get pointer to state and change it. what do you think?

File: beacon-chain/core/helpers/rewards_penalties.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: thanks, done

File: beacon-chain/core/helpers/rewards_penalties.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: done

File: beacon-chain/core/helpers/rewards_penalties.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: done

File: beacon-chain/core/helpers/rewards_penalties.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: done

File: beacon-chain/core/helpers/rewards_penalties.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: thanks

File: beacon-chain/core/helpers/rewards_penalties.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: @terencechain ?

Comment (terencechain) [reply beacon-chain/core/helpers/rewards_penalties.go]: we should return the state there, similar to other helper functions. These are pure functions

Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: thanks

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 6]
```diff
@@ -83,3 +83,69 @@ func InactivityPenalty(
...
+//     """
+//     Decrease the balance for a validator with the given ``index`` by ``delta``.
+//     Set to ``0`` when underflow.
+//     """
+//     current_balance = get_balance(state, index)
+//     set_balance(state, index, current_balance - delta if current_balance >= delta else 0)
+func DecreaseBalance(state *pb.BeaconState, idx uint64, delta uint64) *pb.BeaconState {
+	currentBalance := Balance(state, idx)
+	if currentBalance >= delta {
+		SetBalance(state, idx, currentBalance-delta)
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]: `SetBalance` returns the state here

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 7]
```diff
@@ -83,3 +83,69 @@ func InactivityPenalty(
...
+	} else {
+		SetBalance(state, idx, 0)
```
Comment (terencechain) [inline beacon-chain/core/helpers/rewards_penalties.go]: here too

File: shared/params/config.go [hunk 8]
```diff
@@ -153,6 +153,7 @@ var defaultBeaconConfig = &BeaconChainConfig{
 	MinDepositAmount:           1 * 1e9,
 	MaxDepositAmount:           32 * 1e9,
 	EjectionBalance:            16 * 1e9,
+	HighBalanceIncrement:       1 * 1E9,
```
Comment (terencechain) [inline shared/params/config.go]:
```suggestion
	HighBalanceIncrement:       1 * 1e9,
```

File: beacon-chain/core/helpers/rewards_penalties.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/rewards_penalties.go]: thanks, much nicer that way

**Pull request merged**